File: jason/src/rpc/reconnect_handle.rs
//! Reconnection for [`RpcSession`].

use std::{rc::Weak, time::Duration};

use derive_more::{Display, From};
use tracerr::Traced;

use crate::{
    platform,
    rpc::{BackoffDelayer, RpcSession, SessionError},
    utils::JsCaused,
};

/// Errors occurring in a [`ReconnectHandle`].
#[derive(Clone, Debug, From, Display, JsCaused)]
#[js(error = "platform::Error")]
pub enum ReconnectError {
    /// Some [`SessionError`] has occurred while reconnecting.
    Session(#[js(cause)] SessionError),

    /// [`ReconnectHandle`]'s [`Weak`] pointer is detached.
    #[display(fmt = "ReconnectHandle is in detached state")]
    Detached,
}

/// External handle used to reconnect to a media server when connection is lost.
///
/// This handle will be passed to a `Room.on_connection_loss` callback.
#[derive(Clone)]
pub struct ReconnectHandle(Weak<dyn RpcSession>);

impl ReconnectHandle {
    /// Instantiates new [`ReconnectHandle`] from the given [`RpcSession`]
    /// reference.
    #[inline]
    #[must_use]
    pub fn new(rpc: Weak<dyn RpcSession>) -> Self {
        Self(rpc)
    }

    /// Tries to reconnect after the provided delay in milliseconds.
    ///
    /// If [`RpcSession`] is already reconnecting then new reconnection attempt
    /// won't be performed. Instead, it will wait for the first reconnection
    /// attempt result and use it here.
    ///
    /// # Errors
    ///
    /// With [`ReconnectError::Detached`] if [`Weak`] pointer upgrade fails.
    ///
    /// With [`ReconnectError::Session`] if error while reconnecting has
    /// occurred.
    pub async fn reconnect_with_delay(
        &self,
        delay_ms: u32,
    ) -> Result<(), Traced<ReconnectError>> {
        platform::delay_for(Duration::from_millis(u64::from(delay_ms))).await;

        let rpc = self
            .0
            .upgrade()
            .ok_or_else(|| tracerr::new!(ReconnectError::Detached))?;

        rpc.reconnect().await.map_err(tracerr::map_from_and_wrap!())
    }

    /// Tries to reconnect [`RpcSession`] in a loop with a growing backoff
    /// delay.
    ///
    /// The first attempt will be performed immediately, and the second attempt
    /// will be performed after `starting_delay_ms`.
    ///
    /// Delay between reconnection attempts won't be greater than
    /// `max_delay_ms`.
    ///
    /// After each reconnection attempt, delay between reconnections will be
    /// multiplied by the given `multiplier` until it reaches `max_delay_ms`.
    ///
    /// If `multiplier` is a negative number then it will be considered as
    /// `0.0`. This might cause a busy loop, so it's not recommended.
    ///
    /// Max elapsed time can be limited with an optional `max_elapsed_time_ms`
    /// argument.
    ///
    /// If [`RpcSession`] is already reconnecting then new reconnection attempt
    /// won't be performed. Instead, it will wait for the first reconnection
    /// attempt result and use it here.
    ///
    /// # Errors
    ///
    /// With [`ReconnectError::Detached`] if [`Weak`] pointer upgrade fails.
    ///
    /// With [`ReconnectError::Session`] if error while reconnecting has
    /// occurred and `max_elapsed_time_ms` is provided.
    pub async fn reconnect_with_backoff(
        &self,
        starting_delay_ms: u32,
        multiplier: f64,
        max_delay: u32,
        max_elapsed_time_ms: Option<u32>,
    ) -> Result<(), Traced<ReconnectError>> {
        BackoffDelayer::new(
            Duration::from_millis(starting_delay_ms.into()),
            multiplier,
            Duration::from_millis(max_delay.into()),
            max_elapsed_time_ms.map(|val| Duration::from_millis(val.into())),
        )
        .retry(|| async {
            self.0
                .upgrade()
                .ok_or_else(|| {
                    backoff::Error::Permanent(tracerr::new!(
                        ReconnectError::Detached
                    ))
                })?
                .reconnect()
                .await
                .map_err(tracerr::map_from_and_wrap!())
                .map_err(backoff::Error::Transient)
        })
        .await
    }
}

File: jason/flutter/lib/reconnect_handle.dart
import 'dart:ffi';

import 'ffi/foreign_value.dart';
import 'jason.dart';
import 'util/move_semantic.dart';
import 'util/nullable_pointer.dart';

typedef _free_C = Void Function(Pointer);
typedef _free_Dart = void Function(Pointer);

typedef _reconnect_with_delay_C = Handle Function(Pointer, Int64);
typedef _reconnect_with_delay_Dart = Object Function(Pointer, int);

typedef _reconnect_with_backoff_C = Handle Function(
    Pointer, Int64, Double, Int64, ForeignValue);
typedef _reconnect_with_backoff_Dart = Object Function(
    Pointer, int, double, int, ForeignValue);

final _free = dl.lookupFunction<_free_C, _free_Dart>('ReconnectHandle__free');

final _reconnect_with_delay =
    dl.lookupFunction<_reconnect_with_delay_C, _reconnect_with_delay_Dart>(
        'ReconnectHandle__reconnect_with_delay');

final _reconnect_with_backoff =
    dl.lookupFunction<_reconnect_with_backoff_C, _reconnect_with_backoff_Dart>(
        'ReconnectHandle__reconnect_with_backoff');

/// External handle used to reconnect to a media server when connection is lost.
///
/// This handle is passed to the `RoomHandle.onConnectionLoss()` callback.
class ReconnectHandle {
  /// [Pointer] to the Rust struct backing this object.
  late NullablePointer ptr;

  /// Constructs a new [ReconnectHandle] backed by the Rust struct behind the
  /// provided [Pointer].
  ReconnectHandle(this.ptr);

  /// Tries to reconnect a `Room` after the provided delay in milliseconds.
  ///
  /// If the `Room` is already reconnecting then new reconnection attempt won't
  /// be performed. Instead, it will wait for the first reconnection attempt
  /// result and use it here.
  ///
  /// Throws `RpcClientException` if reconnect attempt fails.
  ///
  /// Throws [StateError] if the underlying [Pointer] has been freed.
  ///
  /// Converts the provided [delayMs] into an `u32`. Throws an [ArgumentError]
  /// if conversion fails.
  Future<void> reconnectWithDelay(int delayMs) async {
    await (_reconnect_with_delay(ptr.getInnerPtr(), delayMs) as Future);
  }

  /// Tries to reconnect a `Room` in a loop with a growing backoff delay.
  ///
  /// The first attempt will be performed immediately, and the second attempt
  /// will be performed after [starting_delay_ms].
  ///
  /// Delay between reconnection attempts won't be greater than [max_delay_ms].
  ///
  /// After each reconnection attempt, delay between reconnections will be
  /// multiplied by the given [multiplier] until it reaches [max_delay_ms].
  ///
  /// If [multiplier] is a negative number then it will be considered as `0.0`.
  /// This might cause a busy loop, so it's not recommended.
  ///
  /// Max elapsed time can be limited with an optional [maxElapsedTimeMs]
  /// argument.
  ///
  /// If the `Room` is already reconnecting then new reconnection attempt won't
  /// be performed. Instead, it will wait for the first reconnection attempt
  /// result and use it here.
  ///
  /// Throws `RpcClientException` if reconnect attempt fails.
  ///
  /// Throws [StateError] if the underlying [Pointer] has been freed.
  ///
  /// Converts the provided [startingDelayMs], [maxDelay] and [maxElapsedTimeMs]
  /// into an `u32`s. Throws an [ArgumentError] if any conversion fails.
  Future<void> reconnectWithBackoff(
      int startingDelayMs, double multiplier, int maxDelay,
      [int? maxElapsedTimeMs]) async {
    var maxElapsedTimeMs_arg = maxElapsedTimeMs == null
        ? ForeignValue.none()
        : ForeignValue.fromInt(maxElapsedTimeMs);

    await (_reconnect_with_backoff(ptr.getInnerPtr(), startingDelayMs,
        multiplier, maxDelay, maxElapsedTimeMs_arg.ref) as Future);
  }

  /// Drops the associated Rust struct and nulls the local [Pointer] to it.
  @moveSemantics
  void free() {
    _free(ptr.getInnerPtr());
    ptr.free();
  }
}
